### PR TITLE
Viper/Ultimate Cleanup

### LINKED ISSUE #1241: Viper Ultimate Wireless not correctly detected in wireless mode
### Distribution

Xubuntu 20.04

### Support

- [x] Have you tried rebooting?
- [x] Have you looked at the [Troubleshooting](https://github.com/openrazer/openrazer/wiki/Troubleshooting) page?

I've installed 2.8.0 and I tried with Polychromatic 0.3.12-stable and Razergenie 0.8.1 with the same result. My Viper Ultimate Wireless is correctly detected and the settings are read correctly only if connected with USB cable, while it is not with the dongle in wireless mode.

Wireless mode
image
image

Wired mode
image
image

### PR DESCRIPTION
**Pull request by nightsky30**
The Viper and Viper Ultimate currently only generate device files and reports for breath, static, and none effects in the driver.

Viper / Viper Ultimate:
 - Cleaned up mouse.py by removing method references to wave effect
 - Removed WAVE_DIRS, HAS_MATRIX, and MATRIX_DIMS as wave wasn't being used (not supported), and they only appear to have Logo LEDs
 - Updated to add support for spectrum and reactive effects
 - Cleaned up fake drivers to reflect the current effect functionality in the driver

Viper Ultimate:
- Added missing battery methods to match driver
- Added missing battery manager
- Fixes #1241 wireless timing issue per [comment](https://github.com/openrazer/openrazer/issues/1241#issuecomment-683392261)

Thanks @ddavini, @Algram, and @troudbal for testing

### CODE REVIEW

**Comment by troudbal** [comment]
As a Viper mouse owner, I can tell that the spectrum and reactive effects are supported. I patched the driver and had no problem https://github.com/troudbal/openrazer/commit/80d55e781c11f8e73378e1f3b44d011ebc187717. As for the wave effect, I can't tell. It's not available in the windows driver.

**Review by z3ntu — COMMENTED** [review]
Tested on all affected devices?

**Comment by nightsky30** [comment]
> Tested on all affected devices?

See https://github.com/openrazer/openrazer/pull/1163/files#r456797607

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 1]
```diff
@@ -1704,15 +1704,10 @@ class RazerViperUltimateWired(__RazerDeviceSpecialBrightnessSuspend):
...
-    MATRIX_DIMS = [1, 20]
-    METHODS = ['get_device_type_mouse', 'max_dpi', 'get_dpi_xy', 'set_dpi_xy', 'get_poll_rate', 'set_poll_rate',
+    METHODS = ['get_device_type_mouse', 'max_dpi', 'get_dpi_xy', 'set_dpi_xy', 'get_poll_rate', 'set_poll_rate', 'get_logo_brightness', 'set_logo_brightness',
                # Logo
-               'get_logo_brightness', 'set_logo_brightness', 'set_logo_wave',
-               'set_logo_static_naga_hex_v2', 'set_logo_spectrum_naga_hex_v2', 'set_logo_none_naga_hex_v2',
-               'set_logo_reactive_naga_hex_v2', 'set_logo_breath_random_naga_hex_v2',
-               'set_logo_breath_single_naga_hex_v2', 'set_logo_breath_dual_naga_hex_v2']
+               'set_logo_static_naga_hex_v2', 'set_logo_spectrum_naga_hex_v2', 'set_logo_none_naga_hex_v2', 'set_logo_reactive_naga_hex_v2',
+               'set_logo_breath_random_naga_hex_v2', 'set_logo_breath_single_naga_hex_v2', 'set_logo_breath_dual_naga_hex_v2', ]
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: remove the comma near the end of the line

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 2]
```diff
@@ -1767,15 +1762,10 @@ class RazerViper(__RazerDeviceSpecialBrightnessSuspend):
...
-    MATRIX_DIMS = [1, 16]
-    METHODS = ['get_device_type_mouse', 'max_dpi', 'get_dpi_xy', 'set_dpi_xy', 'get_poll_rate', 'set_poll_rate',
+    METHODS = ['get_device_type_mouse', 'max_dpi', 'get_dpi_xy', 'set_dpi_xy', 'get_poll_rate', 'set_poll_rate', 'get_logo_brightness', 'set_logo_brightness',
                # Logo
-               'get_logo_brightness', 'set_logo_brightness', 'set_logo_wave',
-               'set_logo_static_naga_hex_v2', 'set_logo_spectrum_naga_hex_v2', 'set_logo_none_naga_hex_v2',
-               'set_logo_reactive_naga_hex_v2', 'set_logo_breath_random_naga_hex_v2',
-               'set_logo_breath_single_naga_hex_v2', 'set_logo_breath_dual_naga_hex_v2']
+               'set_logo_static_naga_hex_v2', 'set_logo_spectrum_naga_hex_v2', 'set_logo_none_naga_hex_v2', 'set_logo_reactive_naga_hex_v2',
+               'set_logo_breath_random_naga_hex_v2', 'set_logo_breath_single_naga_hex_v2', 'set_logo_breath_dual_naga_hex_v2', ]
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: same

**Review by z3ntu — CHANGES_REQUESTED** [review]
minor thing, LGTM otherwise

**Comment by Algram** [comment]
I tested this with my Viper Ultimate and it seems the battery reporting is broken:
image

**Comment by nightsky30** [comment]
@Algram Thanks! I added what was needed for the battery notifier. Can you verify everything is reading well? Battery status, brightness, dpi, polling, firmware, serial? If not, we might need to experiment with different transaction IDs.

**Comment by Algram** [comment]
@nightsky30 I tested it again: When connected to a cable the "getBattery" function returns a value like `8.627450980392156`. This seems correct and it is rising when charging.

When wireless though the getBattery function still returns nothing unfortunately. It would be cool if you could get it to work for the wireless mode as well, since this is where knowing the battery status is most useful.

I think the problem might be that the wireless mouse reports as a completely separate device:
image

On the top there is the viper ultimate with a cable attached and below that with the cable removed. It is also marked as "unknown" with the cable removed.

**Comment by nightsky30** [comment]
@Algram That UNKNOWN can show up even when an entry is added in some cases. Are you seeing any values returned at all? And do you have a frontend to check with like RazerGenie?

The entries for the wireless receiver are already present, so I'm thinking this is a similar issue to what we've seen with the Lancehead Wireless, Mamba Wireless, Atheris, and Basilisk X HyperSpeed in that the mice read values inconsistently due to timing issues. And the wireless receivers need much longer wait times. Otherwise they would return values consistent with what you posted. Serial number would show as UNKNOWN, and firmware, battery, brightness, dpi, polling rate would all show as 0s.

Can you kindly pull the latest and retest? Fingers crossed :smile:

**Comment by Algram** [comment]
@nightsky30 The UNKNOWN is gone and it reports the battery correctly when in wireless mode!

On the downside: In wired mode the battery now reports 0. This worked one commit before.

Thanks!

**Comment by nightsky30** [comment]
Ok, now it should only use the longer wait times for the wireless mode. Let me know if you see any reporting issues. :smile:

**Comment by Algram** [comment]
@nightsky30 Still does not work for the wired mode for some reason..

**Comment by nightsky30** [comment]
> @nightsky30 Still does not work for the wired mode for some reason..

@algram Ok, I think there are some intermittent timing issues with the Viper Ultimate similar to other recent wireless models. The fact that you are reading values SOME of the time is good! But I think the wired battery reporting once and then not a second time even after the timing for the wired device was reset to the original values shows it is only reporting properly part of the time.

I cannot definitively test and set the proper values as I do not own the device. Tweaking the timings is best left to someone who physically has the mouse. They can either do some brute-force testing or perhaps capture packet traffic with wireshark and take a closer look at packet response times.

I'm inclined to open an issue for the wireless receiver for someone else to tackle, and leave that alone with regard to this clean up PR.

Again, the good news is that you are getting responses for battery, serial, etc showing those are working! Thank you for testing :smile:

**Comment by Algram** [comment]
@nightsky30 Thanks for your help! As I own the device I would gladly try to brute force the correct timings. Could you give me a pointer to what pieces of code I would need to change in order to trial and error this?

**Comment by nightsky30** [comment]
@Algram That would be excellent!

You should only need to modify ```driver/razermouse_driver.h``` and ```driver/razermouse_driver.c```:

```driver/razermouse_driver.h```
https://github.com/openrazer/openrazer/blob/6c71ccfc2bbecc014bb13dd301b2b703ef75fcfa/driver/razermouse_driver.h#L68-L74

Here you can see the default WAIT values for mice (in μs):
```
RAZER_MOUSE_WAIT_MIN_US
RAZER_MOUSE_WAIT_MAX_US
```
As well as the special values for the newer wireless mice (also in μs):
```
RAZER_NEW_MOUSE_RECEIVER_WAIT_MIN_US
RAZER_NEW_MOUSE_RECEIVER_WAIT_MAX_US
```
You may need to define new values specifically for the Viper Ultimate if it doesn't work with the \*_NEW_MOUSE_RECEIVER_WAIT_\* values.

```driver/razermouse_driver.c```
https://github.com/openrazer/openrazer/blob/7facc9429ab7e6711776a4eb8b3af520bf6ff0a8/driver/razermouse_driver.c#L43-L61

Modify razer_get_report so that it includes proper case statements for the Viper Ultimate wired and receiver devices. The device IDs for these should already be:

```
USB_DEVICE_ID_RAZER_VIPER_ULTIMATE_WIRED
USB_DEVICE_ID_RAZER_VIPER_ULTIMATE_WIRELESS
```

You may need a separate switch block that calls razer_get_usb_response with WAIT arguments specific to the Viper Ultimate if that device's times are different than those shared by the Mamba Wireless and Lancehead Wireless mice.

When I tested for the Lancehead Wireless, it took a while to get the proper wait values. It was very tedious. It might be more work up front, but easier to test in a VM with USB passthrough so you aren't repeatedly rebooting your bare metal box.

Change the WAIT values in razermouse_driver.h, build, install, reboot, test and see if everything reads properly.

When I tested, I would insert and remove the wireless USB dongle 15-20 times and keep track of how many times in a row settings would read properly. When I could remove and insert the device 15 times with it reliably reading all values, I considered it stable. If it failed at all, then I upped the values and started over (rebuild, reboot, etc).

Sometimes if the timings weren't high enough it might read only half the values reported in the frontend. The serial number might return, and the firmware might show, but perhaps some of the brightness values would display as 0 even though the lights were on. Sometimes the brightness might reflect accurately, but you may be missing the firmware version.

Again, it also might be easier to record packets and simply look at the response times with Wireshark. There's stuff on that in the wiki.

**Comment by z3ntu** [comment]
Is still WIP? (based on the most recent comment)

**Comment by nightsky30** [comment]
> Is still WIP? (based on the most recent comment)

@z3ntu No, @ddavini helped test it in #1241. I consider this done and ready to merge.

**Review by z3ntu — APPROVED** [review]
LGTM, thanks a lot!

**Pull request merged**